modules/interactive_music/audio_stream_playlist.cpp

```diff
@@ -0,0 +1,406 @@
+			double bpm = playlist->audio_streams[idx]->get_bpm();
+			int beat_count = playlist->audio_streams[idx]->get_beat_count();
+			double length;
+			if (bpm > 0.0 && beat_count > 0) {
+				length = beat_count * 60.0 / bpm;
+			} else {
+				length = playlist->audio_streams[idx]->get_length();
+			}
+			if (play_ofs < length) {
+				play_index = i;
+				stream_todo = length - play_ofs;
+				break;
+			} else {
+				play_ofs -= length;
+			}
+		}
+	}
+
+	if (play_index == -1) {
+		return;
+	}
+
+	playback[play_index]->start(play_ofs);
+	fade_index = -1;
+	loop_count = 0;
+
+	active = true;
+}
+
+void AudioStreamPlaybackPlaylist::seek(double p_time) {
+	stop();
+	start(p_time);
+}
+
+int AudioStreamPlaybackPlaylist::mix(AudioFrame *p_buffer, float p_rate_scale, int p_frames) {
+	if (active != true) {
+		for (int i = 0; i < p_frames; i++) {
+			p_buffer[i] = AudioFrame(0.0, 0.0);
+		}
+		return p_frames;
+	}
+
+	double time_dec = (1.0 / AudioServer::get_singleton()->get_mix_rate());
+	double fade_dec = (1.0 / playlist->fade_time) / AudioServer::get_singleton()->get_mix_rate();
+
+	int todo = p_frames;
+
+	while (todo) {
+		int to_mix = MIN(todo, MIX_BUFFER_SIZE);
+
+		playback[play_order[play_index]]->mix(mix_buffer, 1.0, to_mix);
+		if (fade_index != -1) {
+			playback[fade_index]->mix(fade_buffer, 1.0, to_mix);
+		}
+
+		offset += time_dec * to_mix;
+
+		for (int i = 0; i < to_mix; i++) {
+			*p_buffer = mix_buffer[i];
+			stream_todo -= time_dec;
+			if (stream_todo < 0) {
+				// No loop, exit.
+				if (!playlist->loop) {
+					for (int j = i; j < todo - i; j++) {
+						p_buffer[i] = AudioFrame(0, 0);
+					}
+					todo = to_mix;
+					active = false;
+					break;
+				}
+				//find next stream.
+				int prev = play_order[play_index];
+
+				for (int j = 0; j < playlist->stream_count; j++) {
+					play_index++;
+					if (play_index >= playlist->stream_count) {
+						_update_order();
+						play_index = 0;
+						loop_count++;
+						offset = time_dec * (to_mix - i);
+					}
+					if (playback[play_order[play_index]].is_valid()) {
+						break;
+					}
+				}
+
+				if (!playback[play_order[play_index]].is_valid()) {
+					todo = to_mix; // Weird error.
+					active = false;
+					break;
+				}
+
+				bool restart = true;
+				if (prev == play_order[play_index]) {
+					// Went back to the same one, continue loop (if it loops) or restart if it does not.
+					if (playlist->audio_streams[prev]->has_loop()) {
+						restart = false;
+					}
+					fade_index = -1;
+				} else {
+					// Move current mixed data to fade buffer;
+					for (int j = i; j < to_mix; j++) {
+						fade_buffer[j] = mix_buffer[j];
+					}
+
+					fade_index = prev;
+					fade_volume = 1.0;
+				}
+
+				int idx = play_order[play_index];
+
+				if (restart) {
+					playback[idx]->start(0); // No loop, just cold-restart
+					playback[idx]->mix(mix_buffer + i, 1.0, to_mix - i); // Fill rest of mix buffer
+				}
+
+				// Update fade todo
```

Comment:
```suggestion
				// Update fade todo.
```
todo should be written as TODO, but it could be confused with a code TODO. Maybe another word would be better.